### PR TITLE
Snaps documentation

### PR DESCRIPTION
**Pull request by rekmarks**
Adds documentation for Snaps. The Snaps docs consist of:
- An introduction / tutorial (courtesy @mapachurro)
- API documentation
- A development guide (still WIP)

We can merge this PR although the development guide is still unfinished, and finish it in the days after the launch.

Supersedes #289, #330

### CODE REVIEW

File: docs/.vuepress/config.js [hunk 1]
```diff
@@ -6,27 +6,46 @@ module.exports = (_ctx) => ({
...
     ['link', { rel: 'icon', href: `/metamask-fox.svg` }],
     ['link', { rel: 'manifest', href: '/manifest.json' }],
     ['meta', { name: 'theme-color', content: '#3eaf7c' }],
     ['meta', { name: 'apple-mobile-web-app-capable', content: 'yes' }],
-    ['meta', { name: 'apple-mobile-web-app-status-bar-style', content: 'black' }],
-    ['link', { rel: 'apple-touch-icon', href: `/icons/apple-touch-icon-152x152.png` }],
-    ['link', { rel: 'mask-icon', href: '/icons/safari-pinned-tab.svg', color: '#3eaf7c' }],
-    ['meta', { name: 'msapplication-TileImage', content: '/icons/msapplication-icon-144x144.png' }],
-    ['meta', { name: 'msapplication-TileColor', content: '#000000' }]
+    [
```
Comment (Gudahtt) [inline docs/.vuepress/config.js:18]: Lots of churn in this PR. Was `prettier` run on this or something? We can move this to another PR.

File: docs/guide/README.md [hunk 2]
```diff
@@ -41,4 +41,4 @@ We are aware that there are constantly new private blockchains being created. If
 - [Getting Started with Truffle](https://www.trufflesuite.com/docs/truffle/quickstart)
 - [Keeping up with Blockchain Development (from ConsenSys Academy)](https://github.com/ConsenSys-Academy/Blockchain-Developer-Bootcamp/blob/main/docs/S00-intro/L6-keeping-up/index.md)
 - Launch into the Web3 space with [ConsenSys Academy's Blockchain Developer Bootcamp](https://consensys.net/academy/bootcamp/)
-- Explore everything ConsenSys at the [ConsenSys Developer Portal](vhttps://consensys.net/developers/)
```
Comment (Gudahtt) [inline docs/guide/README.md:44]: Not introduced in this PR, not, this looks like a typo. There's a `v` at the start of the URL.

File: docs/guide/rpc-api.md [hunk 3]
```diff
@@ -29,14 +29,14 @@ Important methods from this API include:
...
-## Permissions
+## Restricted Methods
 
 MetaMask introduced Web3 Wallet Permissions via [EIP-2255](https://eips.ethereum.org/EIPS/eip-2255).
-In this permissions system, each RPC method is either _restricted_ or _open_.
-If a method is restricted, an external _domain_ (like a web3 site) must have the corresponding permission in order to call it.
-Open methods, meanwhile, do not require permissions to call, but may require confirmation by the user in order to succeed (e.g. `eth_sendTransaction`).
+In this permissions system, MetaMask is the _host_ environment, and all of its RPC methods are either _restricted_ or _unrestricted_.
+If a method is restricted, an external _subject_ (like a web3 site) must have the corresponding permission in order to call it.
+Unrestricted methods, meanwhile, do not require permissions to call, but may require confirmation by the user in order to succeed (e.g. `eth_sendTransaction`).
```
Comment (Gudahtt) [inline docs/guide/rpc-api.md:37]: `eth_sendTransaction` is not a great example of this because it's a restricted method. It requires `eth_accounts`.

File: docs/guide/snaps-development-guide.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# Snaps Development Guide
+
+::: warning 🛠️ Under Construction 🛠️
+This page is under construction.
+:::
+
```
Comment (Gudahtt) [inline docs/guide/snaps-development-guide.md:6]: I'm guessing that this won't still be under construction when this gets merged

```suggestion
```

Comment (Gudahtt) [reply docs/guide/snaps-development-guide.md:6]: I see multiple references to this being under construction in the text of this page. Hmm. This doesn't seem ready.

We should be able to write a basic development guide that uses our snap template and the snap CLI.

File: docs/guide/snaps-introduction.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
+# Snaps Tutorial
```
Comment (Gudahtt) [inline docs/guide/snaps-introduction.md:1]: This does not seem like a tutorial. Is this a mistake?

File: docs/guide/snaps-introduction.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+## What is Snaps?
+
+Snaps are a new evolution of MetaMask's functionality. By using MetaMask Flask in conjunction with the Snaps API, developers are able to bring their own code to the wallet in a custom-made code laboratory. MetaMask is the first wallet to provide this openness and flexibility to developers, paving the way to scale the number of features and personalization available to the wallet and your Dapp.
```
Comment (Gudahtt) [inline docs/guide/snaps-introduction.md:5]: In the first three sentences, we've described snaps as "a new evolution", and as a "custom-made code laboratory", and congratulated ourselves on being the first to create such a thing, all without saying what it is. This does not seem helpful. We should be explaining what snaps is, not trying to sell it and the wallet itself.

File: docs/guide/snaps-introduction.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+Snaps allows developers to build and plug in their own APIs to MetaMask. No longer do you have to rely on MetaMask to roll out features and updates. The rate of innovation, and MetaMask's flexibility, are limited only by your imagination and your capacity to build.
```
Comment (Gudahtt) [inline docs/guide/snaps-introduction.md:7]: The first sentence is good except it kinda sells snaps short a bit. Building APIs doesn't quite encompass things like enhancements to the wallet itself or integrating with internal MetaMask APIs.

The part about not having to rely on MetaMask to roll out new features and updates doesn't ring true to me. I get what you were going for, that one of the goals of snaps is to allow developers to innovate faster than our team alone could hope to. But maybe we can phrase this more carefully. Snap developers will still very much be dependent upon us and our release schedule.

The last sentence doesn't ring true either :sweat_smile: There remain MANY limitations to what developers can do with snaps. They run in a very restricted environment and the API they are given is quite small right now. Frankly it's probably the most restrictive environment most snap devs will have ever encountered.

File: docs/guide/snaps-introduction.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+## What is Snaps?
+
+Snaps are a new evolution of MetaMask's functionality. By using MetaMask Flask in conjunction with the Snaps API, developers are able to bring their own code to the wallet in a custom-made code laboratory. MetaMask is the first wallet to provide this openness and flexibility to developers, paving the way to scale the number of features and personalization available to the wallet and your Dapp.
+
+Snaps allows developers to build and plug in their own APIs to MetaMask. No longer do you have to rely on MetaMask to roll out features and updates. The rate of innovation, and MetaMask's flexibility, are limited only by your imagination and your capacity to build. 
+
+## MetaMask Snaps: Overview
+
+Broadly speaking, MetaMask Snaps consists of two things: (1) a way to run untrusted JavaScript inside the MetaMask application, and (2) APIs for websites and MetaMask to communicate with individual snaps. As with MetaMask’s [Ethereum JavaScript provider API](https://docs.metamask.io/guide/ethereum-provider.html), communication occurs via JSON-RPC requests and responses.
```
Comment (Gudahtt) [inline docs/guide/snaps-introduction.md:11]: I find the division of snaps into "two things" here as a very odd way to frame things.

I would describe snaps as "a way to run untrusted JavaScript inside the MetaMask application", that part is good. Then maybe we could go on to explain the API that we offer to snaps, and that we offer APis that allow websites to communicate with snaps. I can give this a shot later.

File: docs/guide/snaps-introduction.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+### Current functionality
+
+At present, snaps can make JSON-RPC requests of MetaMask, and websites can make JSON-RPC requests of MetaMask and snaps running inside MetaMask. In this way, snaps can expand MetaMask’s RPC API and change the behavior of the MetaMask application at runtime. Over time, MetaMask will use this to support different blockchains and other decentralized protocols, novel kinds of cryptography, novel kinds of crypto assets, and a countless number of other features.
```
Comment (Gudahtt) [inline docs/guide/snaps-introduction.md:15]: This makes it seem like the ability for snaps to extend the RPC API is what will enable decentralized protocols, novel cryptography, etc. That's not really true. Primarily it will be us building snap APIs that allow customizing the wallet experience that will enable that stuff. Extending the RPC API is just one thing a snap can do among many things.

Also we should probably not be framing "novel kinds of cryptography" as a future feature. We will enable that at launch.

File: docs/guide/README.md (hunk 2 above)
Comment (Gudahtt) [reply docs/guide/README.md:44]: Fixed in #338

File: docs/.vuepress/config.js (hunk 1 above)
Comment (Gudahtt) [reply docs/.vuepress/config.js:18]: I submitted a separate `lint:fix` PR but this file wasn't touched by the linter.

File: docs/guide/snaps-introduction.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+For the prototype Snaps system, snaps cannot modify the MetaMask UI, but can extend the MetaMask RPC API, and exchange arbitrary messages with websites visited by the user. Therefore, the user interface for any snap must exist entirely in the website during the prototype stage.
```
Comment (Gudahtt) [inline docs/guide/snaps-introduction.md:17]: We might want to rephrase this as "Currently snaps cannot modify the MetaMask UI" rather than talking about a "prototype snaps system". I don't know what that could refer to other than Flask, and Flask will definitely support custom UI.

File: docs/guide/snaps-introduction.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+### Execution Environment
+
+As stated, snaps are untrusted JavaScript programs, but execute safely inside the MetaMask application. To isolate snaps from the rest of the application, and to provide a “fully virtualizable” execution environment, MetaMask uses [Secure EcmaScript (SES)](https://github.com/endojs/endo/tree/master/packages/ses), a subset of JavaScript developed by [Agoric](https://agoric.com/). Among other things, SES allows us to restrict access to global JavaScript APIs and to isolate untrusted code from other parts of the application. SES does this at the cost of some performance, and incompatibility with some JavaScript practices, such as modifying prototypes of intrinsic objects, e.g. the Promise constructor.
```
Comment (Gudahtt) [inline docs/guide/snaps-introduction.md:21]: Nit: Reads better with the "As stated" omitted. Readers might be coming directly to this section rather than reading chronologically anyway.

```suggestion
Snaps are untrusted JavaScript programs, but execute safely inside the MetaMask application. To isolate snaps from the rest of the application, and to provide a “fully virtualizable” execution environment, MetaMask uses [Secure EcmaScript (SES)](https://github.com/endojs/endo/tree/master/packages/ses), a subset of JavaScript developed by [Agoric](https://agoric.com/). Among other things, SES allows us to restrict access to global JavaScript APIs and to isolate untrusted code from other parts of the application. SES does this at the cost of some performance, and incompatibility with some JavaScript practices, such as modifying prototypes of intrinsic objects, e.g. the Promise constructor.
```

File: docs/guide/snaps-introduction.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+* nodejs `^14.15`
+* nvm (recommended)
+* git
+* yarn v1 
+* macOS, Linux, or Windows Subsystem Linux
+  
+### Getting Flask running in your local environment
+
+Flask, as described [here](LINK TO FLASK PAGE REQUIRED HERE), is a distribution channel for developmental, experimental, kooky stuff for MetaMask, and Snaps is the first feature to be released through it.
```
Comment (Gudahtt) [inline docs/guide/snaps-introduction.md:37]: We should update this description to match the latest draft, this "kooky stuff" one was replaced.

File: docs/guide/snaps-introduction.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+* nvm (recommended)
+* git
+* yarn v1 
+* macOS, Linux, or Windows Subsystem Linux
+  
+### Getting Flask running in your local environment
+
+Flask, as described [here](LINK TO FLASK PAGE REQUIRED HERE), is a distribution channel for developmental, experimental, kooky stuff for MetaMask, and Snaps is the first feature to be released through it. 
+
+* Start by downloading [Flask](LINK HERE) and adding it to Chrome
```
Comment (Gudahtt) [inline docs/guide/snaps-introduction.md:39]: We have the link now!
```suggestion
* Start by downloading [Flask](https://chrome.google.com/webstore/detail/metamask-flask-developmen/ljfoeinjpaedjfecbmggjgodbgkmjkjk) and adding it to Chrome
```

File: docs/guide/snaps-introduction.md (hunk 5 above)
Comment (Gudahtt) [reply docs/guide/snaps-introduction.md:1]: Since this page is named "Snaps Introduction" and there is a subheader below called "Snaps Tutorial", this should probably be:

```suggestion
# Snaps Introduction
```

**Comment by Gudahtt** [comment]
Just rebased this on `main`

File: docs/.vuepress/config.js [hunk 14]
```diff
@@ -6,27 +6,46 @@ module.exports = (_ctx) => ({
...
+    ],
+    ['meta', { name: 'msapplication-TileColor', content: '#000000' }],
   ],
 
   theme: '@vuepress/theme-default',
 
   themeConfig: {
     repo: 'MetaMask/metamask-docs',
-    docsDir: 'docs',
+    docsDir: 'packages/docs/dist',
```
[the remark this answers is not in the document]
Comment (Gudahtt) [reply docs/.vuepress/config.js:48]:
```suggestion
    docsDir: 'docs',
```

File: docs/guide/rpc-api.md (hunk 3 above)
Comment (Gudahtt) [reply docs/guide/rpc-api.md:37]:
```suggestion
Unrestricted methods, meanwhile, do not require permissions to call, but may require confirmation by the user in order to succeed (e.g. `wallet_addEthereumChain`).
```

Comment (Gudahtt) [reply docs/guide/rpc-api.md:37]: Addressed in https://github.com/MetaMask/metamask-docs/pull/342

File: docs/.vuepress/config.js (hunk 1 above)
Comment (Gudahtt) [reply docs/.vuepress/config.js:18]: Addressed here: https://github.com/MetaMask/metamask-docs/pull/346

File: docs/guide/rpc-api.md [hunk 15]
```diff
@@ -29,14 +29,14 @@ Important methods from this API include:
...
 - [`eth_sign`](https://eth.wiki/json-rpc/API#eth_sign)
 
-## Permissions
+## Restricted Methods
 
 MetaMask introduced Web3 Wallet Permissions via [EIP-2255](https://eips.ethereum.org/EIPS/eip-2255).
-In this permissions system, each RPC method is either _restricted_ or _open_.
-If a method is restricted, an external _domain_ (like a web3 site) must have the corresponding permission in order to call it.
-Open methods, meanwhile, do not require permissions to call, but may require confirmation by the user in order to succeed (e.g. `wallet_addEthereumChain`).
+In this permissions system, MetaMask is the _host_ environment, and all of its RPC methods are either _restricted_ or _unrestricted_.
```
Comment (Gudahtt) [inline docs/guide/rpc-api.md:35]: I extracted these changes to https://github.com/MetaMask/metamask-docs/pull/348

I left out the part about MetaMask being the "host" and callers being "subjects" though. In this context those terms are confusing, since they aren't fully explained and are never used again.

I also removed the link to the implementation altogether rather than updating it.

File: docs/guide/snaps-rpc-api.md [hunk 16]
```diff
@@ -0,0 +1,579 @@
...
+
+[[toc]]
+
+## Unrestricted Methods
+
+### `wallet_enable`
+
+::: warning Only Callable By
+
+- Websites
```
Comment (Gudahtt) [inline docs/guide/snaps-rpc-api.md:21]: I'm surprised to see these methods here. I was thinking of the "Snaps API" as "the API provided to snaps". The idea being that this would be one cohesive API.  I would expect the snap-related methods offered to websites to be in our main RPC API docs.

Comment (rekmarks) [reply docs/guide/snaps-rpc-api.md:21]: Right, and currently it's "Snaps API" as in "the API related to snaps, regardless of to whom / what it's exposed". I don't know that the website-only methods map neatly on the existing JSON-RPC API page either, in that snaps can call many / some of those methods as well (although perhaps they shouldn't be able to by default).

Comment (Gudahtt) [reply docs/guide/snaps-rpc-api.md:21]: We can explain the subset of the RPC API that snaps get access to in the Snaps reference documentation. I think that would be the simplest way to organize this. The RPC API docs are somewhat of a mess right now because of the feature parity issue between mobile and extension, but at least aspirationally we want to converge at some point on a shared API that is standardized between all products. Since dapps will rely on these snap-related methods, they belong there.

File: docs/guide/snaps-development-guide.md (hunk 4 above)
Comment (Gudahtt) [reply docs/guide/snaps-development-guide.md:6]: The "snaps tutorial" part of the intro could be adapted into a decent "development guide". It's not far off.

File: docs/guide/snaps-introduction.md (hunk 12 above)
Comment (Gudahtt) [reply docs/guide/snaps-introduction.md]: This is the blurb I was thinking of:

>Flask is a distribution channel, or playground, for developers that gives them access to additional unstable APIs. The goal of Flask is to maximize developer control, so that we can learn the full extent of what developers want to do with MetaMask, and may consider incorporating those lessons into the main MetaMask distribution.

File: docs/guide/snaps-introduction.md (hunk 6 above)
Comment (mapachurro) [reply docs/guide/snaps-introduction.md]: So, what do you propose? Do we cut this copy entirely?

Comment (Gudahtt) [reply docs/guide/snaps-introduction.md]: This is what I have so far for this section:

>Snaps is a system that allows safely expanding the capabilities MetaMask. A "snap" is a program that we run in an isolated environment that can customize the wallet experience in some way. For example, a snap can add new APIs to MetaMask, add new features, or modify existing functionality using internal APIs. Snaps makes extending MetaMask easier and safer, for both users and developers.

File: docs/guide/snaps-introduction.md (hunk 8 above)
Comment (Gudahtt) [reply docs/guide/snaps-introduction.md]: On further consideration I think it was just the specific wording here that was throwing me off. Generally this isn't bad.

Here's what I came up with trying to re-write it. Not sure that it's better though.

>Broadly speaking, MetaMask Snaps is a system that allows running untrusted programs inside MetaMask and granting them specific capabilities. Snaps are run in an isolated environment where they have access to a limited set of capabilities, determined by the permissions they were granted by the user during installation. As with MetaMask’s [Ethereum JavaScript provider API](https://docs.metamask.io/guide/ethereum-provider.html), snaps communicate with MetaMask using a JSON-RPC API.
>
>New RPC methods have been added to our [Ethereum JavaScript provider API](https://docs.metamask.io/guide/ethereum-provider.html) as well. These methods allow websites to request that snaps be installed, and to communicate with individual snaps.

File: docs/guide/snaps-introduction.md (hunk 9 above)
Comment (Gudahtt) [reply docs/guide/snaps-introduction.md]: Rather than updating this to be consistent with the reference docs, maybe we can delete this entire section and refer there instead?

Comment (Gudahtt) [reply docs/guide/snaps-introduction.md]: Or if we want to keep it, here's my current draft:

>At present, snaps can create new RPC methods for websites to call, they can call many of the same RPC methods that websites can call, and they have access to a limited set of snap-exclusive RPC methods (see the Snaps RPC API documentation for more information). Over time, MetaMask will expose more internal functionality as RPC methods, granting more capabililities to snaps.

File: docs/guide/snaps-introduction.md (hunk 10 above)
Comment (Gudahtt) [reply docs/guide/snaps-introduction.md]: Here's my current draft:

>At the moment, snaps cannot modify the MetaMask UI directly. If a snap needs a UI, that UI must exist entirely on a website that can communicate with the snap. We will be allowing UI customization in future updates.

File: docs/guide/snaps-development-guide.md (hunk 4 above)
Comment (mapachurro) [reply docs/guide/snaps-development-guide.md:6]: That sounds like what I intended from the beginning, so--a rose by any other name!

File: docs/guide/snaps-introduction.md (hunk 12 above)
Comment (mapachurro) [reply docs/guide/snaps-introduction.md]: Yeah, that's fine--I had thought we had gotten rid of the 'kooky stuff' copy throughout the resources, despite, you know, loving it :-P

File: docs/guide/snaps-rpc-api.md (hunk 16 above)
Comment (mapachurro) [reply docs/guide/snaps-rpc-api.md:21]: Is this something that needs to be resolved before tomorrow morning?

File: docs/guide/snaps-introduction.md (hunk 6 above)
Comment (mapachurro) [reply docs/guide/snaps-introduction.md]: I'm OK with this if you guys are.

File: docs/guide/snaps-introduction.md (hunk 7 above)
Comment (mapachurro) [reply docs/guide/snaps-introduction.md]: -I wasn't sure how much of that we were going to push, and how soon; hence I left it deliberately vague
-There was copy you found elsewhere--about how we'll use Flask to get a better idea of what devs want, and incorporate those things faster--that might resolve your doubts here.
-How about we nix that last sentence, then?

File: docs/guide/snaps-introduction.md (hunk 8 above)
Comment (mapachurro) [reply docs/guide/snaps-introduction.md]: > On further consideration I think it was just the specific wording here that was throwing me off. 

I have this experience a lot when collaborating on a text, so I feel you

Your version gives a lot more detail, and given that we're writing for devs here to explain how to do something complex, that seems appropriate to me.

File: docs/guide/snaps-introduction.md (hunk 9 above)
Comment (mapachurro) [reply docs/guide/snaps-introduction.md]: I think deleting and referring the users directly to the Snaps API section of the MM RPC API docs sounds sensible. No need to frame things differently in different places.

That is what you mean by 'refer there instead', right?

File: docs/guide/snaps-introduction.md (hunk 10 above)
Comment (mapachurro) [reply docs/guide/snaps-introduction.md]: That's fine--again, I was downplaying the UI modification capabilities intentionally.

**Review by mapachurro — APPROVED** [review]
To Snaps I say: "come up to the lab... And see what's on the slab"

File: docs/guide/snaps-introduction.md (hunk 7 above)
Comment (Gudahtt) [reply docs/guide/snaps-introduction.md]: Sorry for the fractured conversation - was thinking that [the draft here](https://github.com/MetaMask/metamask-docs/pull/331#discussion_r786326845) could replace both of these paragraphs. Do you think we need anything from this one?

I feel like there is still _something_ to say about the rate of innovation but I couldn't think of how to say it. Maybe the Flask description gets close to it. It's not an exact parallel because Flask is basically just for allowing experimentation and gaining feedback to improve the snaps system, whereas with snaps the ultimate goal isn't for all of the good ideas to get absorbed into MetaMask itself. The good ideas can stay as snaps, that's fine. But there is a parallel in that the systems are both meant to drive innovation, and they work together toward that goal.

But I don't know how we should say any of that here, and it seems fine if we omit it for now.

File: docs/guide/snaps-introduction.md (hunk 9 above)
Comment (Gudahtt) [reply docs/guide/snaps-introduction.md]: Yep! OK.

File: docs/guide/snaps-rpc-api.md (hunk 16 above)
Comment (Gudahtt) [reply docs/guide/snaps-rpc-api.md:21]: It's important that we resolve it soon, for the sake of starting snaps off with a solid foundation in terms of the structure of the documentation. But I don't think it's strictly needed for tomorrow morning, no

**Review by Gudahtt — APPROVED** [review]
